### PR TITLE
Add a StorageInfo API

### PR DESCRIPTION
**Pull request by asutula**
- Adds a "query" api to `Scheduler.cistore` to get all or some list of `StorageInfo`
- Uses that api in FFS to provide similar functionality for an FFS instance
- Exposes that api via the client API and CLI

Want to get quick feedback on that and then will add similar apis to the `admin` side of things.

### CODE REVIEW

File: ffs/api/api.go [hunk 1]
```diff
@@ -130,19 +130,6 @@ func (i *API) GetStorageConfigs(cids ...cid.Cid) (map[cid.Cid]ffs.StorageConfig,
 	return configs, nil
 }
 
-// Show returns the information about a stored Cid. If no information is available,
-// since the Cid was never stored, it returns ErrNotFound.
-func (i *API) Show(cid cid.Cid) (ffs.StorageInfo, error) {
```
Comment (asutula) [inline ffs/api/api.go:135]: Moved this and gave it a new/better name.

File: ffs/api/api_storageinfo.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+	"fmt"
+
+	"github.com/ipfs/go-cid"
+	"github.com/textileio/powergate/ffs"
+	"github.com/textileio/powergate/ffs/scheduler"
+)
+
+// StorageInfo returns the information about a stored Cid. If no information is available,
+// since the Cid was never stored, it returns ErrNotFound.
+func (i *API) StorageInfo(cid cid.Cid) (ffs.StorageInfo, error) {
```
Comment (asutula) [inline ffs/api/api_storageinfo.go:13]: New/better name for `Show`

File: ffs/api/api_storageinfo.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+		return inf, ErrNotFound
+	}
+	if err != nil {
+		return inf, fmt.Errorf("getting cid storage info: %s", err)
+	}
+	return inf, nil
+}
+
+// QueryStorageInfo returns a list of infomration about all stored cids, filtered by cids if provided.
+func (i *API) QueryStorageInfo(cids ...cid.Cid) ([]ffs.StorageInfo, error) {
```
Comment (asutula) [inline ffs/api/api_storageinfo.go:25]: New FFS func to list all or some of the `StorageInfo` for the FFS instance

File: ffs/scheduler/internal/cistore/cistore.go [hunk 4]
```diff
@@ -45,6 +51,67 @@ func (s *Store) Get(iid ffs.APIID, c cid.Cid) (ffs.StorageInfo, error) {
 	return ci, nil
 }
 
+// Query returns a list of StorageInfo matching any provided query options.
+func (s *Store) Query(iids []ffs.APIID, cids []cid.Cid) ([]ffs.StorageInfo, error) {
```
Comment (asutula) [inline ffs/scheduler/internal/cistore/cistore.go:55]: New `cistore` method that lists `StorageInfos` for all/some instances and/or cids. Tried to be efficient about how it is implemented without going as far as making indexes.

File: ffs/scheduler/scheduler_storage.go [hunk 5]
```diff
@@ -105,6 +105,11 @@ func (s *Scheduler) GetStorageInfo(iid ffs.APIID, c cid.Cid) (ffs.StorageInfo, e
 	return info, nil
 }
 
+// QueryStorageInfo returns a list of StorageInfo matching any provided query options.
+func (s *Scheduler) QueryStorageInfo(iids []ffs.APIID, cids []cid.Cid) ([]ffs.StorageInfo, error) {
```
Comment (asutula) [inline ffs/scheduler/scheduler_storage.go:109]: Just bubbling the new `cistore` method up into the scheduler layer. This is used by the FFS API, and if the general api looks good, the `Admin` gRPC api will call this too.

File: api/client/storageinfo.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+	client userPb.UserServiceClient
+}
+
+// StorageInfo returns the information about a stored Cid. If no information is available,
+// since the Cid was never stored, it returns an error with codes.NotFound.
+func (s *StorageInfo) StorageInfo(ctx context.Context, cid string) (*userPb.StorageInfoResponse, error) {
+	return s.client.StorageInfo(ctx, &userPb.StorageInfoRequest{Cid: cid})
+}
+
+// QueryStorageInfo returns a list of infomration about all stored cids, filtered by cids if provided.
```
Comment (jsign) [inline api/client/storageinfo.go:20]: Nit: typo.

File: api/server/user/storageinfo.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+		return nil, status.Errorf(codes.NotFound, "getting storage info: %v", err)
+	}
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "getting storage info: %v", err)
+	}
+	pbInfo := toRPCStorageInfo(info)
+	return &userPb.StorageInfoResponse{StorageInfo: pbInfo}, nil
+}
+
+// QueryStorageInfo returns a list of infomration about all stored cids, filtered by cids if provided.
```
Comment (jsign) [inline api/server/user/storageinfo.go:36]: Nit: typo.

File: cli-docs/pow/pow_storage-info_list.md [hunk 8]
```diff
@@ -0,0 +1,29 @@
+## pow storage-info list
```
Comment (jsign) [inline cli-docs/pow/pow_storage-info_list.md:1]: Nit: Might be good to rename to `query` to be in sympathy with the API? (or rename `Query` to `List`).
I think both are fine, just mentioning they both use different names.

File: cmd/pow/cmd/storage_info_get.go [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+func init() {
+	storageInfoCmd.AddCommand(storageInfoGetCmd)
+}
+
+// Cmd is the command.
```
Comment (jsign) [inline cmd/pow/cmd/storage_info_get.go:16]: Nit: Intentional comment?

File: cmd/pow/cmd/storage_info_list.go [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+func init() {
+	storageInfoCmd.AddCommand(storageInfoListCmd)
+}
+
+var storageInfoListCmd = &cobra.Command{
+	Use:   "list [optional cid1,cid2,...]",
+	Short: "Returns a list of infomration about all stored cids, filtered by cids if provided.",
```
Comment (jsign) [inline cmd/pow/cmd/storage_info_list.go:19]: Nit: typo.

File: ffs/scheduler/scheduler_storage.go (hunk 5 above)
Comment (jsign) [reply ffs/scheduler/scheduler_storage.go:109]: Nit: Mind wrapping the error to have some more path-explanation of it for this layer?

**Review by jsign — APPROVED** [review]
LGTM!

File: cli-docs/pow/pow_storage-info_list.md (hunk 8 above)
Comment (asutula) [reply cli-docs/pow/pow_storage-info_list.md:1]: Made it `List`

**Pull request merged**